domain/consensus/processes/reachabilitymanager/reachability_stretch_test.go

```diff
@@ -47,89 +53,203 @@ func runJsonDAGTest(t *testing.T, fileName, testName string, addArbitraryBlocks,
 	}
 	defer gzipReader.Close()
 
-	err = tc.MineJSON(gzipReader)
+	tips, err := tc.MineJSON(gzipReader)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = tc.ReachabilityManager().ValidateIntervals(tc.DAGParams().GenesisHash)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	err = tc.ReachabilityManager().ValidateIntervals(params.GenesisHash)
+	return tips
+}
+
+func addArbitraryBlocks(t *testing.T, tc testapi.TestConsensus) {
+	// After loading json, add arbitrary blocks all over the DAG to stretch
+	// reindex logic, and validate intervals post each addition
+
+	blocks, err := tc.ReachabilityManager().GetAllNodes(tc.DAGParams().GenesisHash)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	if addArbitraryBlocks {
-		// After loading json, add arbitrary blocks all over the DAG to stretch reindex logic
-		// and validate intervals post each addition
+	numChainsToAdd := len(blocks) / 2 // Multiply the size of the DAG with arbitrary blocks
+	maxBlocksInChain := 20
+	validationFreq := int(math.Max(1, float64(numChainsToAdd/100)))
+
+	randSource := rand.New(rand.NewSource(33233))
+
+	for i := 0; i < numChainsToAdd; i++ {
+		randomIndex := randSource.Intn(len(blocks))
+		randomParent := blocks[randomIndex]
+		newBlock, _, err := tc.AddUTXOInvalidHeader([]*externalapi.DomainHash{randomParent})
+		if err != nil {
+			t.Fatal(err)
+		}
+		blocks = append(blocks, newBlock)
+		// Add a random-length chain every few blocks
+		if randSource.Intn(8) == 0 {
+			numBlocksInChain := randSource.Intn(maxBlocksInChain)
+			chainBlock := newBlock
+			for j := 0; j < numBlocksInChain; j++ {
+				chainBlock, _, err = tc.AddUTXOInvalidHeader([]*externalapi.DomainHash{chainBlock})
+				if err != nil {
+					t.Fatal(err)
+				}
+				blocks = append(blocks, chainBlock)
+			}
+		}
+		// Normally, validate intervals for new chain only
+		validationRoot := newBlock
+		// However every 'validation frequency' blocks validate intervals for entire DAG
+		if i%validationFreq == 0 || i == numChainsToAdd-1 {
+			validationRoot = tc.DAGParams().GenesisHash
+		}
+		err = tc.ReachabilityManager().ValidateIntervals(validationRoot)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func addAlternatingReorgBlocks(t *testing.T, tc testapi.TestConsensus, tips []*externalapi.DomainHash) {
+	// Create alternating reorgs to test the cases where
+	// reindex root is out of current virtual's chain
```

Comment:
It's not virtual, because virtual is only for block with bodies. The more accurate chain is `the header selected tip chain`